### PR TITLE
Use EngineNetworkStream to feed instances of ShoutOutput

### PR DESCRIPTION
**Pull request by Palakis**
This PR removes EngineBroadcast's thread and FIFO buffer from the new multi-broadcasting subsystem, and lets EngineNetworkStream directly feed the FIFO buffer of each ShoutOutput instance.

It works as is, but there are two design issues left to settle:
- In EngineNetworkStream, the frame count is used to properly synchronize with the parent SoundDeviceNetwork object through `EngineNetworkStream::getWriteExpected()`. What is the proper way to refactor this call to take the several underlying workers into account?
- What about the input parts of `EngineNetworkStream`?

### CODE REVIEW

File: src/engine/sidechain/enginenetworkstream.cpp [hunk 1]
```diff
@@ -90,81 +106,104 @@ int EngineNetworkStream::getReadExpected() {
...
-    }
-    int copyCount = math_min(writeAvailable, writeRequired);
-    if (copyCount > 0) {
-        (void)m_pOutputFifo->write(buffer, copyCount);
-        // we advance the frame only by the samples we have actually copied
-        // This means in case of buffer full (where we loose some frames)
-        // we do not get out of sync, and the syncing code tries to catch up the
-        // stream by writing silence, once the buffer is free.
-        m_streamFramesWritten += copyCount / m_numOutputChannels;
+        if(!worker->threadWaiting()) {
```
Comment (daschuer) [inline src/engine/sidechain/enginenetworkstream.cpp]: Nit: please add a blank after "if"

File: src/engine/sidechain/enginenetworkstream.h [hunk 2]
```diff
@@ -48,7 +50,8 @@ class EngineNetworkStream {
     qint64 m_streamStartTimeUs;
     qint64 m_streamFramesWritten;
     qint64 m_streamFramesRead;
-    QSharedPointer<NetworkStreamWorker> m_pWorker;
+
+    QList<QSharedPointer<NetworkStreamWorker>> m_workers;
```
Comment (daschuer) [inline src/engine/sidechain/enginenetworkstream.h]: I am afraid this construct is not thread save to be filled in on thread and looped in the engine thread. 
I think we can user a QVector with a reasonable number of predefined fields. 
Locking by a mutex or so is no option in the engine thread.

Comment (daschuer) [inline src/engine/sidechain/enginenetworkstream.h]: On a the one thought: Is the list touched at runtime? Will enabling and disabling outputs touch the list? 
QShared pointer is t

**Comment by daschuer** [comment]
.. thread save. So you can fill the original QList with a reasonable amount of pointers and treat it as const list from that point.
What can be the maximum parallel stream you can think of? 4?

**Comment by Palakis** [comment]
> What can be the maximum parallel stream you can think of? 4?
For me, 16 would be a good number. Large enough for most use cases.

File: src/engine/sidechain/enginenetworkstream.h (hunk 2 above)
Comment (Palakis) [inline src/engine/sidechain/enginenetworkstream.h]: > I think we can user a QVector with a reasonable number of predefined fields.
Locking by a mutex or so is no option in the engine thread.

Understood. 

> Is the list touched at runtime? Will enabling and disabling outputs touch the list?

The list is touched only when adding or removing a connection in BroadcastSettings.

File: src/engine/sidechain/enginebroadcast.cpp [hunk 3]
```diff
@@ -118,10 +105,14 @@ bool EngineBroadcast::removeConnection(BroadcastProfilePtr profile) {
     if(!profile)
         return false;
 
+    m_connectionsMutex.lock();
```
Comment (daschuer) [inline src/engine/sidechain/enginebroadcast.cpp]: Is removeConnection and addConnection called from a different thread? 
Please add comments to the source about the locking conditions.

File: src/broadcast/defs_broadcast.h [hunk 4]
```diff
@@ -26,4 +26,6 @@
 #define BROADCAST_FORMAT_MP3 "MP3"
 #define BROADCAST_FORMAT_OV "OggVorbis"
 
+#define BROADCAST_MAX_CONNECTIONS 16
```
Comment (daschuer) [inline src/broadcast/defs_broadcast.h]: Please add a comment to source, why this is not dynamic.

File: src/engine/sidechain/enginenetworkstream.cpp [hunk 5]
```diff
@@ -40,6 +41,8 @@ EngineNetworkStream::EngineNetworkStream(int numOutputChannels,
       m_streamFramesWritten(0),
       m_streamFramesRead(0),
       m_writeOverflowCount(0) {
+    m_workers.reserve(BROADCAST_MAX_CONNECTIONS);
```
Comment (daschuer) [inline src/engine/sidechain/enginenetworkstream.cpp]: QVector is not guaranteed to be thread save, but QSharedPointer is. So you can here add all 16 QSharedPointers and use reset() to add and remove a worker.

File: src/preferences/dialog/dlgprefbroadcast.cpp [hunk 6]
```diff
@@ -220,6 +220,13 @@ void DlgPrefBroadcast::enableCustomMetadataChanged(int value) {
 }
 
 void DlgPrefBroadcast::btnCreateConnectionClicked(bool enabled) {
+    if(m_pBroadcastSettings->rowCount() >= BROADCAST_MAX_CONNECTIONS) {
+        QMessageBox::warning(this, tr("Action forbidden."),
```
Comment (daschuer) [inline src/preferences/dialog/dlgprefbroadcast.cpp]: "Action failed" sounds more gentle.

File: src/engine/sidechain/enginebroadcast.cpp (hunk 3 above)
Comment (Palakis) [inline src/engine/sidechain/enginebroadcast.cpp]: These are supposedly called outside the engine thread. I'll check that to be sure and add a comment about this.

File: src/engine/sidechain/enginenetworkstream.cpp (hunk 5 above)
Comment (Palakis) [inline src/engine/sidechain/enginenetworkstream.cpp]: Okay so my initial approach of using QVector like a QList is not right.

File: src/engine/sidechain/enginenetworkstream.cpp [hunk 7]
```diff
@@ -90,81 +110,104 @@ int EngineNetworkStream::getReadExpected() {
...
+                worker->incrementOverflowCount();
+            }
+
+            int copyCount = math_min(writeAvailable, writeRequired);
+            if (copyCount > 0) {
+                (void)workerFifo->write(buffer, copyCount);
+                // we advance the frame only by the samples we have actually copied
+                // This means in case of buffer full (where we loose some frames)
+                // we do not get out of sync, and the syncing code tries to catch up the
+                // stream by writing silence, once the buffer is free.
```
Comment (Palakis) [inline src/engine/sidechain/enginenetworkstream.cpp]: Seems like SoundDeviceNetwork is responsible for writing silence. Is it reasonable to try reproducing this in EngineNetworkStream?

File: src/engine/sidechain/enginenetworkstream.cpp [hunk 8]
```diff
@@ -90,81 +110,104 @@ int EngineNetworkStream::getReadExpected() {
...
+                worker->addFramesWritten(copyCount / m_numOutputChannels);
```
Comment (Palakis) [inline src/engine/sidechain/enginenetworkstream.cpp]: Seems like SoundDeviceNetwork is responsible for writing silence. Is it reasonable to try reproducing this in EngineNetworkStream?

File: src/engine/sidechain/enginebroadcast.cpp (hunk 3 above)
Comment (Palakis) [inline src/engine/sidechain/enginebroadcast.cpp]:
```
Debug [Main]: EngineNetworkStream::addWorker: worker added
Debug [Main]: EngineNetworkStream: worker slots used: "3 out of 16"
Debug [Main]: EngineBroadcast::addConnection: created connection for profile "Profile 3"
Debug [Main]: EngineNetworkStream::removeWorker: worker removed
Debug [Main]: EngineNetworkStream: worker slots used: "2 out of 16"
Debug [Main]: EngineBroadcast::removeConnection: removed connection for profile "Profile 3"
```

These are called outside the engine thread.

**Comment by Palakis** [comment]
So I managed to have proper buffer management. To achieve this, SoundDeviceNetwork is now aware of the several streaming connections, and most of the worker and buffer management code has been moved to NetworkStreamWorker, which has more logic and responsibilities than before.
I may need to remove EngineNetworkStream and move its worker list to SoundDeviceNetwork and its Input part into a separate class.

**Pull request merged**